man1/ydate.1
.\" Automatically generated by Pod::Man v1.37, Pod::Parser v1.35
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sh \" Subsection heading
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  | will give a
.\" real vertical bar.  \*(C+ will give a nicer C++.  Capital omega is used to
.\" do unbreakable dashes and therefore won't be available.  \*(C` and \*(C'
.\" expand to `' in nroff, nothing in troff, for use with C<>.
.tr \(*W-|\(bv\*(Tr
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.Sh), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.if \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.\"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.hy 0
.if n .na
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "YDATE 1"
.TH YDATE 1 "2010-12-16" "perl v5.8.9" "User Contributed Perl Documentation"
.SH "NAME"
.Vb 1
\& ydate - Calculate a date in the timeline by taking +-N days as parameter
.Ve
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
.Vb 1
\& ydate [ -d ] | [ --days=+|-N ] | --help
.Ve
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
.Vb 4
\& Calculate a date in the timeline by taking +-N days as parameter
\& Mandatory arguments:
\&       -d, --days=+N|-N                by default -1 day, otherwise +-N days
\&       -h, --help                      show help
.Ve
.PP
.Vb 1
\& Report bugs to <petrohs(at) mail.com>
.Ve
.SH "EXAMPLES"
.IX Header "EXAMPLES"
.Vb 2
\& andresaquino $> date "+%Y%m%d"
\& 20101216
.Ve
.PP
.Vb 2
\& andresaquino $> sh ydate.sh -d
\& 20101215
.Ve
.PP
.Vb 2
\& andresaquino $> sh ydate.sh --days=-4
\& 20101212
.Ve
.PP
.Vb 2
\& andresaquino $> sh ydate.sh --days=5
\& 20101221
.Ve
.SH "AUTHOR"
.IX Header "AUTHOR"
.Vb 3
\& Authors
\&   Carlos Aquiles < (at) mail.com>  
\&   PetrOHS <petrohs(at) mail.com>
.Ve
.PP
.Vb 2
\& Maintainer
\&   Andres Aquino <andres.aquino(at) mail.com>
.Ve
.SH "BUGS"
.IX Header "BUGS"


ydate.sh
#!/bin/sh 
# vim: set ts=2 sw=2 sts=2 et si ai: 

# ydate.sh
# =-=
# This software is Free Software; it can be redistributed under the same terms as BSD Licence.
# All rights reserved.
#
# Authors
#   Carlos Aquiles < (at) mail.com>  
#   PetrOHS <petrohs(at) mail.com>
# Maintainer
#   Andres Aquino <andres.aquino(at) mail.com>
# 

# getNumberDay
# Get a number day taking the Gregorian Calendar as a baseline
# Params:
#   getNumberDay( integer year,
#                 integer month,
#                 integer day )
# Returns:
#   integer NumberDay
#
function getNumberDay () {
  local year=${1}
  local month=${2}
  local day=${3}
  local number=
  
  # values in range year > 0 || month > 0 || day > 0 => exit 1
  [ ${year} -le 0 -o ${month} -le 0 -o ${day} -le 0 ] && exit 1

  month=$(((month+9)%12))
  year=$((year-(month/10)))
  number=$(((365*year) + (year/4) - (year/100) + (year/400) + (month*306 + 5)/10 + (day -1)))

  echo $number
}

# getDayOfNumber
# Get a date from a NumberDay generated by getNumberDay function
# Params:
#   getDayOfNumber ( integer numberDay )
# Returns:
#   string dateCalculated
# 
function getDayOfNumber () {
  local number=${1}
  local year=
  local month=
  local day=
 
  year=$(((10000*number+14780)/3652425))
  day=$((number-((365*year)+(year/4)-(year/100)+(year/400))))
  if [ ${day} -lt 0 ]
  then
    year=$((year-1))
    day=$((number-((365*year)+(year/4)-(year/100)+(year/400))))
  fi
  ydays=$((((100*day)+52)/3060))
  month=$((((ydays+2)%12)+1))
  year=$((year+(ydays+2)/12))
  day=$((day-((ydays*306+5)/10)+1))
  
  # formatting
  month=`echo "0000${month}" | rev | cut -c-2 | rev`
  day=`echo "0000${day}" | rev | cut -c-2 | rev`
  echo "${year}${month}${day}"

}

# sdate.sh 
# Params:
#   -d            by default -1 day
#   --days=-+N    move N days back or forward
#   -h | --help   show help
#
while [ $# -gt 0 ]
do
  case ${1} in
    -d)
      # opcion por defecto (-1)
      days=-1
      ;;
   --days| --days=*)
      # corrimiento de dias (+-N)
      days=`echo "${1}" | sed 's/^--days[=]*//'`
      [ ${#days} -eq 0 ] && days=-1 
      ;;
    -h|--help)
      echo -e "Usage: ydate [OPTIONS]..."
      echo -e "Calculate a date in the timeline by taking +-N days as parameter"
      echo -e "Mandatory arguments:"
      echo -e "\t-d, --days=+N|-N                by default -1 day, otherwise +-N days"
      echo -e "\t-h, --help                      show help"
      echo -e "Report bugs to <petrohs(at) mail.com>"
      exit 0
      ;;
    *)
      # opcion por defecto (-1)
      echo -e "Option not recognized, please check your parameters."
      echo -e "ydate --help"
      exit 0
      ;;
  esac
  shift
done

# get current date
nyear=`date '+%Y'`
nmonth=`date '+%m'`
nday=`date '+%d'`

# get number day
numberday=$(getNumberDay "${nyear}" "${nmonth}" "${nday}")

# apply n shift days
numberday=$((numberday+(days)))

# and finally, get a date
getDayOfNumber "${numberday}"


# "Compartir es bueno"
exit 0;

